Add console tree helper

othercorey opened the pull request:
refs https://github.com/cakephp/cakephp/issues/18632

~Lists are printed as values without keys unless the value is an array.~

othercorey committed:
Add console tree helper

dereuromark commented:
I tested it with the linked symfony reference from https://symfony.com/blog/new-in-symfony-7-3-new-and-improved-console-helpers
and dataset
```php
$data = [
	'src/' =>  [
		'Command/',
		'Controller/' => [
			'DefaultController.php',
		],
		'Kernel.php',
	],
	'templates/' => [
		'base.html.twig',
	],
];

$io->helper('Tree')->output($data);
```

I got
```
src/
├── 0
│   └── Command/
├── Controller/
│   └── DefaultController.php
└── 1
    └── Kernel.php
templates/
└── base.html.twig
```

But probably their demo result output is more in line what one would expect:

```
├── src/
│   ├── Command/
│   ├── Controller/
│   │   └── DefaultController.php
│   └── Kernel.php
└── templates/
    └── base.html.twig
```

dereuromark commented:
We should probably focus on the values, and especially their string representation if possible.

Further tests with "stringish objects" yield following results

```
		Html::create('<b>x</b>'),
		UserStatus::Active,
		...
```
printed as
```
├── 2
│   └── (Templating\View\Html)
└── 3
    └── (Sandbox\Model\Enum\UserStatus)
```
In their string form those would be ideally printed as
```
│   ├── <b>x</b>
│   └── Active    
```

The first requires a string cast or maybe some lookup for Stringable.
The 2nd the similar Enum checks and label() calls we have in other places.

Maybe also callables and other ways of wrapping stringish output should work.
Currently those render as (Closure) etc.

Anything that cannot translate to string should probably rather issue a warning/error instead of silently printing out.
People could always manually cast to whatever they need if the defaults are not sufficient.

I made a PR for us: https://github.com/dereuromark/cakephp-sandbox/pull/92

dereuromark commented:
View TreeHelper so far also didnt have it yet though
https://github.com/dereuromark/cakephp-tools/pull/300
Made some tests, I think Closure checks should suffice, we probably dont have to support any callable.

ADmad commented on src/Command/Helper/TreeHelper.php:
```diff
@@ -0,0 +1,89 @@
+     */
+    protected function outputValue(mixed $value, string $prefix): void
+    {
+        if (is_scalar($value)) {
```
```suggestion
        if (is_scalar($value) || $value instanceof Stringable) {
```

ADmad commented on src/Command/Helper/TreeHelper.php:
```diff
@@ -0,0 +1,89 @@
+        $i = 1;
+        $isList = array_is_list($array);
+        foreach ($array as $key => $value) {
+            $isLast = $i++ === count($array);
```
The `count()` call should be moved out of the loop.

othercorey replied:
True, but I had hoped it was optimized to a single value lookup anyway.

othercorey commented:
Agree that Stringable  objects should be printed. I left it minimal after getting a functional layout to get comments.

othercorey commented:
To think of this more like a file system listing, changed to only show key for string keys or array values instead of for any non-list.

Added warning message for non-stringable tree nodes.

dereuromark commented:
Please also see my other comments, specifically the enum part. See the linked tree Helper changes.
I think those should be included to support the common inputs.

othercorey commented:
> Please also see my other comments, specifically the enum part. See the linked tree Helper changes. I think those should be included to support the common inputs.

Interesting. If you think people will create trees with closures in them, we can support it.

othercorey commented:
If we add options like initial indent, should that be part of the TreeHelper instance config or does that need to be an option passed to `output()` where we don't assume $args is the tree itself?

dereuromark commented:
That sounds more like Helper config directly

othercorey commented:
- Added support for `initialIndent` config
- Added support for closures returning strings
- Added support for enum values
- Fixed display of bool values

Dropped warning for non-stringable values. Users can deal with anything that's invalid by looking at the output.

Was able to consolidate the top-level and nested array element output so testing value output from one is valid for all.

dereuromark commented on src/Command/Helper/TreeHelper.php:
```diff
@@ -0,0 +1,117 @@
+    {
+        if ($value instanceof Closure) {
+            $this->_io->out($prefix . $value());
+        } elseif ($value instanceof UnitEnum) {
```
This could be enhanced to have a more convenient output for labels if possible IMO.
See my linked PR

So `5` currently would be displayed as `Active`.

dereuromark replied:
if ($content instanceof EnumLabelInterface) {
					$content = $content->label();
				} elseif ($content instanceof BackedEnum) {
					$content = $content->value;
				} elseif ($content instanceof UnitEnum) {
					$content = $content->name;
				} 
				
not sure if we need the difference between backed and unit though.

othercorey replied:
I understand the idea but what is this tree we're trying to display?

dereuromark replied:
If you dont want to add it now thats fine, I can make follow up PRs.
The whole point of this is to not have to extend and customize in the first place :) A helper that helps instead of blocks people from displaying values I think would be the idea.

dereuromark replied:
For this specific example imagine one wants to display the users persmissions/roles.
It would be more cumbersome to have to correctly format than just passing the enums and know its gonna be formatted according to Cake principles.

othercorey replied:
Added EnumLabelInterface and BackedEnum cases.

dereuromark reviewed (approved):
Tested it and looks really great.

Apart from the one small enum topic I think this seems fine to be merged.
